### DISCUSSION TITLE
https tps 측정시 성능 문제

### DISCUSSION THREAD

**Discussion by mk778-song** [naver/ngrinder#961] [Q&A]
안녕하세요. 

aws를 이용하여 구성된 서버에서  aws cloudfront에 https request를 전송하여 tps를 측정했는데, 
https의 tps가 http 대비 2~4배 정도 느리게 측정되네요.

기존에 javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure 오류가 발생해서
jsse.enableSNIExtension 값을 true로 설정해놓은 상태입니다.

tps가 낮게 나오는게 http2 문제인가 싶어서 HttpVersionPolicy.FORCE_HTTP_2 로 설정하고 테스트 해보니 
org.apache.hc.core5.http.ConnectionClosedException: Connection is closed 오류가 발생합니다.

근데 또 curl 로 cloudfront에 http2 request를 보내보면 정상적으로 동작하네요.

도움 주시면 감사하겠습니다.

### 실행환경
jdk 1.8
controller : aws ec2 / tomcat 8.5
agent : aws ecs

**Comment by junoyoon** [comment]
https 는 초기에 handshaking 이 일어나기 때문에, 2~4배 느리게 나올 수 있습니다.
왠만하면 http 로 테스트 하기 바랍니다.

@imbyungjun 님 handshaking 을 어느 시점에 하죠?

**Reply by mk778-song** [reply]
답변 감사합니다.

그런데 HTTPPluginControl.getConnectionDefaults().setFollowRedirects(true) 옵션은 안되는 걸까요??

보안 때문에 되도록이면 http 포트는 닫아놓고, http를 https로 redirect해서 테스트해보려는데 301 또는 307 코드가 나오네요

https://github.com/naver/ngrinder/issues/280 참고해서 @BeforeThread 에 적용하였습니다.

**Reply by imbyungjun** [reply]
현재 nGrinder에는 두 가지 HTTP 클라이언트가 존재합니다. 간단하게 구 HTTP 클라이언트 / 신규 HTTP 클라이언트라고 지칭하겠습니다.
글에 적어주신 내용을 보았을때는 신규 HTTP 클라이언트를 사용하신 것 같습니다. 신규 HTTP 클라이언트에는 현재 redirect 옵션은 없습니다.
`HTTPPluginControl.getConnectionDefaults().setFollowRedirects(true)`옵션은 구 HTTP 클라이언트에 적용되는 옵션입니다.
스크립트 생성시 "Groovy (Legacy)"를 선택하여 구 HTTP 클라이언트를 사용하도록 하면 redirect 가능할 것 같습니다. 다만, 구 HTTP 클라이언트는 HTTP/2.0 통신을 맺을 수 없습니다.

**Comment by junoyoon** [comment]
@imbyungjun could u check this?

**Comment by donggyu04** [comment]
@mk778-song 

님 신규 httpclient 사용하시고 아래 옵션 선택 해제하시고 테스트 한번 부탁드립니다.
아마 성능이  나쁘지 않게 나올거에요

![image](https://user-images.githubusercontent.com/14273601/227081255-96c8627f-d3b7-4ebd-ae03-83efa18b6605.png)
